website/blog/2023-02-10.ferretdb-compare-arrays.md

```diff
@@ -0,0 +1,605 @@
+  { _id: 'nested-number', v: [ [ 6, 1 ] ] },       // [ 6, 1 ] is used for comparison.
+  { _id: 'nested-mixed', v: [ [ true, 0, 'g' ] ] } // [ true, 0, 'g' ] is used for comparison.
+]
+```
+
+When comparing two nested arrays, they do not iterate each index.
+Instead, it gets the smallest inner array and uses that to compare with other array.
+
+For example, if comparison iterated each index of `nested-null` and `nested-number`,
+the first index of the nested array `'g'` of `nested-null` 
+would be greater than `6` of `nested-number`. The fact `nested-null` is smaller than `nested-number`
+indicates that, even nested array to nested array comparison uses the smallest element from an array
+for comparison.
+
+Let's look at how `nested-null` and `nested-number` are compared.
+
+1. `nested-null` is an array, get the smallest element `[ null ]`.
+2. `nested-number` is an array, get the smallest element `[ 6, 1 ]`.
+3. Compare `[ null ]` and `[ 6, 1 ]` using BSON comparison order, both are arrays.
+4. Compare the first indexes of the array using BSON comparison order,
+   `null` is null type and less than number type `6`. 
+
+So `nested-null` is less than `nested-number`.
+
+For a nested array to array comparison, BSON comparison order is used.
+Let's look at how `array-mixed` and `nested-empty` are compared.
+1. `array-mixed` is an array, get the smallest element `0`.
+2. `nested-empty` is an array, get the smallest element `[]`.
+3. Compare `0` and `[]` using BSON comparison order, `0` is number and less than array `[]`.
+
+So `array-mixed` is less than `nested-empty`.
+
+## Comparison in Filter Operator
+
+[$gt and $lt](https://docs.ferretdb.io/basic_operations/read/#retrieve-documents-using-operator-queries)
+filter operators select records that are greater than or less than the query.
+
+Filtering used in `$lt` and `$gt` operators use following rules
+to compare array and other BSON types.
+
+* `$lt` uses the smallest element of the same BSON type as non-array from the array.
+* `$gt` uses the largest element of the same BSON type as non-array from the array.
+* Empty array should not match other BSON type.
+
+Let's insert array records to apply `$gt` and `$lt` filters.
+
+```js
+db.mixedfilter.insertMany([
+  { _id: 'array-empty', v: [] },
+  { _id: 'array-mixed', v: [ 0, 'a', 'z' ] },
+  { _id: 'array-string', v: [ 'd' ] },
+  { _id: 'array-number', v: [ 10, 5 ] },
+  { _id: 'string', v: 'b' },
+  { _id: 'null', v: null }
+])
+```
+
+`$lt` selects records that are less than a specified value but this is only applied to the same BSON type. 
+Let's look at filtering less than `'c'`.
+
+```js
+db.mixedfilter.find({v: {$lt: 'c'}})
+[
+  { _id: 'array-mixed', v: [ 0, 'a', 'z' ] }, // 'a' the smallest string is used for comparison.
+  { _id: 'string', v: 'b' }                   // 'b' is used for comparison.
+]
+```
+The records selected are string BSON type `string` and array BSON type `array-mixed`.
```

Comment:
```suggestion
```

The records selected are string BSON type `string` and array BSON type `array-mixed`.
```